### DISCUSSION TITLE
Help with testing jogging using keyboard shortcuts

### DISCUSSION THREAD

**Discussion by breiler** [winder/Universal-G-Code-Sender#2164] [General]
Hi, 

I would appreciate some help testing jogging with keyboard shortcuts. We have had an old bug where keyboard shortcuts for jogging would get stuck (#1494). 

I have now added a special shortcut handler that will check if the action supports a continuous action that will act on key press and release. This way when a keyboard shortcut for jogging is released it will send a jog cancel and it should stop almost immediately.

This is available in the latest nightly build https://github.com/winder/Universal-G-Code-Sender#downloads

The shortcuts will now have two behaviours. 
* If you press the shortcut and release within 500ms it will do a single step using the step size setting.
* If you press and hold the shortcut it will continously jog with the set feed speed and should stop jogging as soon as you release the keys

Any feedback is appreciated!

**Comment by highlandparkfurniture** [comment]
Hey, I recently had to reconfigure all my keymaps and changed from using a gamepad to using a joystick mapped to a keyboard with AntiMicro. I noticed the changes you've made and everything is working as you described. This update couldn't have come at a better time, thank you! I'll spend some time adjusting the settings in AntiMicro to the new UGS jogging behavior, and I'll reply again if I figure out how to optimize things for anyone else using that program. 

Would it be possible to add diagonal motion if two keys are pressed?

**Reply by breiler** [reply]
No, unfortunatley you can't combine keyboard actions for diagonal movement. For that you would need to use UGS native support for gamepads/joysticks.

**Reply by highlandparkfurniture** [reply]
I ended up using JoyToKey to map my joystick to a virtual number pad(8,2,4,6), and was able to map the diagonals to 7,9,1,3. This made it very easy to assign these to diagonal jog commands in UGS. It is definetly not as smooth as the native jogging support, but if someone is limited in their setup they could work with it.

Some feedback on the jogging bahavior:

It would be nice to be able to adjust or deactivate the press/hold delay timing in the settings. I found the 500ms delay before the machine starts moving to be a little sluggish. I typically use the GUI jog controller for  for deliberate or precise step-sized movements, and keys/shortcuts when I'm eyeballing it. I see the appeal behind the two behaviors, but I personally would rather a short button press equals a short jog. When trying to make a fine adjustment you might suddenly send your tool across the whole work piece.

I found it a bit clunky when transitioning between jog directions: machine will jog X+ until you hit Y+, it will then slow down and stop for half a second before moving Y+. Smoother cornering would be nice, if it's possible.

Sort of a combination of the other 2 notes, I noticed if I randomly hit a bunch of keys, the machine sort of just works its way through the queue. Like if I just mash the numb pad the machine travels in short random bursts, stopping for a half second then going of in a new direction. When I take my hand off the controls the machine keeps moving for several seconds. Obviosly, I was just testing it out and no one should be spamming the controls on their machine, but maybe you can send a jog cancel before transitioning from single-step commands to continuous jog commands?
